media/sse/src/main/java/org/glassfish/jersey/media/sse/internal/JerseyEventSink.java

```diff
@@ -118,15 +158,38 @@ public void flush() throws IOException {
         super.flushQueue();
     }
 
-    public void onComplete() {
-        checkClosed();
-        subscription.cancel();
-        close();
+    @Override
+    protected void onClose(Exception e) {
+        cancelSubscription();
+    }
+
+    private void cancelSubscription() {
+        if (subscription != null) {
+            subscription.cancel();
+        }
     }
 
     private void checkClosed() {
         if (isClosed()) {
+            cancelSubscription();
             throw new IllegalStateException(LocalizationMessages.EVENT_SOURCE_ALREADY_CLOSED());
         }
     }
+
+    private MediaType resolveMediaType(Object item) {
+        // all stream is presumed to be of items with same type
+        if (implicitMediaType == null) {
+            Class<?> clazz = item.getClass();
+            if (Number.class.isAssignableFrom(clazz)
+                    || Character.class.isAssignableFrom(clazz)
```

Comment:
Character, String and Boolean are final classes. Faster is just to compare them.

Reply:
My bad sorry